analyses/chromosomal-instability/util/chr-break-calculate.R

```diff
@@ -0,0 +1,233 @@
+    ranges = IRanges::IRanges(
+      start = dplyr::pull(break_df, !!rlang::sym(start_col)),
+      end = dplyr::pull(break_df, !!rlang::sym(end_col))
+    ),
+    mcols = dplyr::pull(break_df, !!rlang::sym(samples_col))
+  )
+
+  return(granges)
+}
+
+break_density <- function(sv_breaks = NULL,
+                          cnv_breaks = NULL,
+                          sample_id = NULL,
+                          window_size = 1e6,
+                          chr_sizes_list = NULL,
+                          samples_col_cnv = "samples",
+                          chrom_col_cnv = "chrom",
+                          start_col_cnv = "start",
+                          end_col_cnv = "end",
+                          samples_col_sv = "samples",
+                          chrom_col_sv = "chrom",
+                          start_col_sv = "start",
+                          end_col_sv = "end") {
+  # For given breaks data.frame(s), calculate the breaks density for a tiled
+  # windows across the genome. Returns the data as a GenomicRanges object for
+  # easy mapping with `ggbio`. Where the density and counts are stored in
+  # @elementMetadata@listData.
+  #
+  # Args:
+  #   sv_breaks: a data.frame with the breaks for the SV data.
+  #   cnv_breaks: a data.frame with the breaks for the SV data.
+  #   sample_id: sample ID for which the data needs to be extracted and made into
+  #              a GenomicRanges object. If "all" is designated, all samples will
+  #              be kept. Multiple samples can be designated as a character vector.
+  #   chr_size_list: a named numeric vector of the sizes (bp) of the chromosomes.
+  #                  names of the chromosomes must match the format of the input
+  #                  break data. e.g. "chr1" or "1".
+  #   window_size: What size windows to calculate break density. Default is 1kb.
+  #   samples_col_sv/cnv: character string that indicates the column name with the
+  #                sample ID information. Default is "samples". Will be passed to
+  #                `make_granges`` function.
+  #   chrom_col_sv/cnv: character string that indicates the column name with the
+  #              chromosome information. Default is "chrom". Will be passed to
+  #              `make_granges`` function.
+  #   start_col_sv/cnv: character string that indicates the column name with the
+  #              start coordinate. Default is "start". Will be passed to
+  #              `make_granges`` function.
+  #   end_col_sv/cnv: character string that indicates the column name with the
+  #            end coordinate. Default is "end". Will be passed to
+  #            `make_granges`` function.
+  #
+  # Check that a sample ID has been specified.
+  if (is.null(sample_id)) {
+    stop("No sample ID has been specified. Use the `sample_id` argument.")
+  }
+
+  # Determine how many samples are in the group
+  n_samples <- length(sample_id)
+
+  # Make CNV into GRanges
+  if (!is.null(cnv_breaks)) {
```

Comment:
I _think_ this has to be changed to `if (!is.null(cnv_breaks) & is.null(sv_breaks))` and same for the next if. Otherwise, we have logic:
1. if not null CNV
2. if not null SV
3. if neither is null <- ends up being duplicative, right? you note this in your comments about overwriting. **Actually** it will probably be most clean to just rearrange: Make the last `if` come first (neither is null), and then some `else if`'s for the other two. Then a final else to `stop()` if both are null.

Reply:
Awesome. I knew I didn't like my logic set up here but wasn't sure how to make it better. This is great.

Reply:
If I'm understanding your suggestion right after another read, I don't think it will work. I can't have the last if come first, because the last if uses the items created in the first two ifs.

Reply:
Basically if you just supply one of the datasets, that will be the only dataset used, but if you supply both datasets, we need to combine them.

Reply:
What I was thinking was this concept - you don't think it will work? 
if (both present)
else if (only CV)
else if(only SV)
else (everything is null)

Also, apologies, I can't test this in depth right now since data release 12 is actively downloading onto work computer that I haven't used in a month, but will look more closely interacting with the code tomorrow!! So if my suggestion makes no sense this is probably why.

Reply:
Let me play around with a bit and see if I can do what you are suggesting here. I definitely like the logic of it better.

Reply:
It just involves some more involved ifs, checking "if one of interest is not null **AND** other is null". The double negatives get trippy here, may be a good task for tomorrow morning..